### ISSUE TITLE
not recognizing id

### ISSUE THREAD

**Issue by BjoernRave** [FormidableLabs/urql-exchange-graphcache#106]
For my whole app it is working great, but for this one page it keeps telling me that its needs an id to cache the request and keeps resending the query. I am querying for an id on every subfield and also verified that my backend is sending those. I though it might be connected to arras, that are requested This is it: 

```
const GET_PLACE = gql`
  query GET_PLACE($id: ID!) {
    place(where: { id: $id }) {
      id
      title
      description
      createdAt
      updatedAt
      isTemporary
      startTime
      endTime
      categories {
        id
        urban
        remote
        activity
        viewpoint
        concert
        fleaMarket
        foodMarket
        artwork
        streetFestival
        artsy
        charity
      }
      accessibility
      cleanliness
      crowdiness
      view
      relaxing
      visits
      accessibilityCount
      cleanlinessCount
      crowdinessCount
      viewCount
      relaxingCount
      location {
        id
        lat
        lng
      }
      images {
        id
        original
      }
      comments {
        id
        text
        createdAt
        author {
          id
          name
        }
      }
      author {
        id
      }
    }
  }
`
```

and this the schema: 
```
type Place {
  id: ID!
  createdAt: DateTime!
  updatedAt: DateTime!
  title: String!
  description: String
  location: GeoJson!
  author: User
  comments(
    where: CommentWhereInput
    orderBy: CommentOrderByInput
    skip: Int
    after: String
    before: String
    first: Int
    last: Int
  ): [Comment!]
  categories: Categories
  ratings(
    where: RatingWhereInput
    orderBy: RatingOrderByInput
    skip: Int
    after: String
    before: String
    first: Int
    last: Int
  ): [Rating!]
  images(
    where: ImageWhereInput
    orderBy: ImageOrderByInput
    skip: Int
    after: String
    before: String
    first: Int
    last: Int
  ): [Image!]
  isTemporary: Boolean!
  startTime: DateTime
  endTime: DateTime
  visits: Int
  accessibility: Float
  view: Float
  cleanliness: Float
  crowdiness: Float
  accessibilityCount: Int
  viewCount: Int
  cleanlinessCount: Int
  crowdinessCount: Int
  relaxing: Float
  relaxingCount: Int
  source: String!
  posted: Boolean
  interaction: Float
}
```

**Comment by kitten** [comment]
Hiya 👋 do you have the specific output from the warning that you can post here? Maybe it’s just a little unclear and we’ll have to improve it?

**Comment by BjoernRave** [comment]
Hey :) 

of course, here it is: 

```
Invalid key: The GraphQL query at the field at `Query.place({"where":{"id":"ck16vgnou0130079207sveo64"}})` has a selection set, but no key could be generated for the data at this field.
You have to request `id` or `_id` fields for all selection sets or create a custom `keys` config for `Place`.
Entities without keys will be embedded directly on the parent entity. If this is intentional, create a `keys` config for `Place` that always returns null.
https://github.com/FormidableLabs/urql-exchange-graphcache/blob/master/docs/help.md#15
```

**Comment by kitten** [comment]
Ok, that seems rather odd, since the query you've posted does include `id` as a field, which is required (`ID!`). Are you sure that all your queries do include this field, just to be thorough 😅

Maybe we should include the query name in the warning in the future

**Comment by lukasluecke** [comment]
Yeah it might be another query that also queries the place, where the `id` selection is missing - maybe you could check your network requests and post the results?

@kitten More context in the warning would be awesome - I ran into the same issue in the past I believe, and that would probably have helped solve it faster.

**Comment by kitten** [comment]
This will probably improve the quality of our warnings and errors by adding a small trace to them. Maybe this helps? It's published for testing as `1.0.4-debug-trace` under the `@urql/exchange-graphcache@debug-trace` tag.

**Comment by BjoernRave** [comment]
@kitten thanks for that :) I found the problematic query, it was indeed as @lukasluecke said. It was this one: 

```
const GET_USER_RATING = gql`
  query GET_RATINGS($userId: ID, $placeId: ID!) {
    place(where: { id: $placeId }) {
      ratings(where: { author: { id: $userId } }) {
        id
        visited
        accessibility
        view
        relaxing
        crowdiness
        cleanliness
      }
    }
  }
`
```
when it should be like this: 

```
const GET_USER_RATING = gql`
  query GET_RATINGS($userId: ID, $placeId: ID!) {
	id
    place(where: { id: $placeId }) {
      ratings(where: { author: { id: $userId } }) {
        id
        visited
        accessibility
        view
        relaxing
        crowdiness
        cleanliness
      }
    }
  }
`
```

I think the improved warning message should be included in the master branch
